design-documents/services-orchestration.md

```diff
@@ -0,0 +1,254 @@
+# Distributed services orchestration by using workflow (state machine)
+
+This is an extension to [Service Isolation](service-isolation.md) document.
+
+## Actual challenges
+Keeping of strict boundaries between services rises a new challenge on how to organize complex cross-system communication that was inherited from the Magento application. The fact sub-applications can be deployed separately lead us to an obvious conclusion that such communication will be performed through the network and this communication does not have a state that can be shared between different components. The old scenario which required interaction across multiple domains, for instance, catalog and checkout, with this model of application deployment will have to invoke such operations remotely. The purpose of this document is to propose a way how to organize and manage such scenarios.
+![Figure 1](services-orchestration/so1.png)
+
+The main question, which must be addressed prior to many others is who will be responsible for organizing such communication? As soon as one service is starting to call other services through the network it becomes responsible for handling responses, managing timeouts, retries etc. As the result, increases overall complexity. Since this moment this service becomes responsible for handling network communication and receives knowledge about of called service like knowledge about foreign service signature. As a result, the service becomes more fragile. Changes that were made at another end may lead to original service invalidation.
+
+Another drawback which should be considered, when one service encapsulates calls to another service such situation may affect the service replaceability. For instance, we have service `AddProductToCart` which encapsulate call to service `GetProductCustomizableOptions`. We can imagine a situation when a developer who knows about service `AddProductToCart` implementation details may build an extension with an assumption that `GetProductCustomizableOptions` will be always called inside of `AddProductToCart`. Replacing the service `AddProductToCart` by service `MyAddProductToCart` which is doing the same but does not call `GetProductCustomizableOptions` may unintentionally break logic of such extension.
+
+Possible "accidental" performance degradation also is a point of consideration. In case we make it possible and convenient to call one service from another service we could loose track of cases when it happens. Network call must not be convenient to place it anywhere in the code (For instance, in a plug-in of a frequently called method). 
```

Comment:
I think this could be an issue in both cases, when services know about each other and when communication between services orchestrated. I would propose to remove this statement.